### ISSUE TITLE
BUG: `[ci skip]` ignored in PR merge

### ISSUE THREAD

**Issue by jakirkham** [travis-ci/travis-ci#6669]
If a merged PR commit has `[ci skip]` in it's commit message, Travis CI seems to skip building that commit, but instead builds the latest commit in the PR on the base branch. This is pretty odd behavior as one would normally expect nothing to be built in this case. In fact other CIs (e.g. AppVeyor or CircleCI), do exactly that (i.e. build nothing).

As an example, consider this PR ( https://github.com/conda-forge/poliastro-feedstock/pull/2 ). It states in the PR title to `[ci skip]`. As expected, all CIs build the commits in the PR. Here's [Travis CI's build]( https://travis-ci.org/conda-forge/poliastro-feedstock/builds/164528051 ). When merging the PR, we get this merge commit ( https://github.com/conda-forge/poliastro-feedstock/commit/af5c402c262c54a27bcde0a1cec634e01ba845b7 ). It can be seen clearly that the commit message says `[ci skip]` in it. All of the CIs skip building that commit.

However, Travis CI starts a [build]( https://travis-ci.org/conda-forge/poliastro-feedstock/builds/164566480 ) of commit ( https://github.com/conda-forge/poliastro-feedstock/commit/950d375e3da41d1b37acacc764c41954c67dc253 ), which was the last commit in the PR. If we look at the status of that [commit]( https://github.com/conda-forge/poliastro-feedstock/commits/950d375e3da41d1b37acacc764c41954c67dc253 ). We can see it has push and pr statuses from Travis CI when it should only have a pr status.

In short, it would be nice when merging a PR that has `[ci skip]` in the merge commit to have nothing build not a re-build of some previous PR commit.

**Comment by BanzaiMan** [comment]
PR title does not affect whether or not we build it. The only commit in https://github.com/conda-forge/poliastro-feedstock/pull/2 is https://github.com/conda-forge/poliastro-feedstock/commit/950d375e3da41d1b37acacc764c41954c67dc253, which does not have any special command to skip CI.

**Comment by jakirkham** [comment]
Sorry I think this issue is being misunderstood. Unfortunately it is a difficult one to explain. I'll try to be more terse and see if that helps.

Commit ( https://github.com/conda-forge/poliastro-feedstock/commit/950d375e3da41d1b37acacc764c41954c67dc253 ) should be built as part of the PR, but it should not be built again as part of a push when merging the PR. Travis CI is building it for both cases hence there are 2 Travis CI statuses.

**Comment by BanzaiMan** [comment]
The merge commit of https://github.com/conda-forge/poliastro-feedstock/pull/2 is https://github.com/conda-forge/poliastro-feedstock/commit/af5c402c262c54a27bcde0a1cec634e01ba845b7. We did not build it.

There are two statuses for the commit because the commit itself does not have `[skip ci]` message in it.

**Comment by jakirkham** [comment]
The only time that normally happens is when someone does the PR from one branch on the repo to another branch on the same repo like PR ( https://github.com/jakirkham/metawrap/pull/1 ). However, if it is done from a fork, like PR ( https://github.com/nanshe-org/nanshe/pull/414 ), then we only get the PR status and not the push status (which is normally reserved for the merge commit). In the case of PR ( https://github.com/conda-forge/poliastro-feedstock/pull/2 ), this was from a fork so there shouldn't be a push status.

Also note that typically when one does have the push and PR build statuses on a commit they tend to be closer together temporally like PR ( https://github.com/jakirkham/metawrap/pull/1 )'s [PR build]( https://travis-ci.org/jakirkham/metawrap/builds/164524476 ) and the [push build]( https://travis-ci.org/jakirkham/metawrap/builds/164524325 ). However, in the case of PR ( https://github.com/conda-forge/poliastro-feedstock/pull/2 ), the [PR build]( https://travis-ci.org/conda-forge/poliastro-feedstock/builds/164528051 ) and the [push build]( https://travis-ci.org/conda-forge/poliastro-feedstock/builds/164566480 ) occurred ~6hrs apart. The latter only occurring after the merge.

In other words, this suggests that on the merge. Travis CI saw the `[ci skip]` in the [merge commit message]( https://github.com/conda-forge/poliastro-feedstock/commit/af5c402c262c54a27bcde0a1cec634e01ba845b7 ). However, instead of just skipping the push build, it looked to the parent commit ( https://github.com/conda-forge/poliastro-feedstock/commit/950d375e3da41d1b37acacc764c41954c67dc253 ) and built that instead. This doesn't seem like desirable behavior to me. When I add `[ci skip]` to the last of a string of commits before pushing, I don't expect it to try and build my second to last commit. Instead I expect it to respect `[ci skip]` in the commit I pushed and not look elsewhere.

**Comment by BanzaiMan** [comment]
I see that there is a bug; we saw the commit https://github.com/conda-forge/poliastro-feedstock/commit/950d375e3da41d1b37acacc764c41954c67dc253 coming in for conda-forge/poliastro-feedstock#2 from a fork, but the Push status was posted for `conda-forge/poliastro-feedstock`, rather than the fork. This is especially dubious, because the fork (conda-forge-admin/poliastro-feedstock) doesn't have Travis CI enabled.

I suspect that the 6-hour lag between the PR build and push build of the merge is probably just a matter of the conda-forge organization's size. It has many Mac builds, and its build queue can be quite long at times, and frankly, it is nearly impossible to predict when certain jobs run. As incredible as it may sound, I think that https://travis-ci.org/conda-forge/poliastro-feedstock/builds/164528051 and https://travis-ci.org/conda-forge/poliastro-feedstock/builds/164566480 probably ran 6 hours apart just as they are intended. Right now, at almost midnight US Eastern time, there are 15 Mac jobs (with 2 running) for the `conda-forge` organization.

**Comment by BanzaiMan** [comment]
I will open a separate issue for the bug I mentioned above, rather than reopening this.

**Comment by jakirkham** [comment]
> ...the Push status was posted for `conda-forge/poliastro-feedstock`, rather than the fork. This is especially dubious, because the fork (conda-forge-admin/poliastro-feedstock) doesn't have Travis CI enabled.

Correct, the fork has [Travis CI disabled]( https://travis-ci.org/conda-forge-admin/poliastro-feedstock/ ). So there should be no push build whatsoever.

> I suspect that the 6-hour lag between the PR build and push build of the merge is probably just a matter of the conda-forge organization's size.

That would be a reasonable theory except for the fact that the [push build]( https://travis-ci.org/conda-forge/poliastro-feedstock/builds/164566480 ) appears on `master` instead of `feedstock_rerender_master`, which is the branch the PR came from. Though if you are still unconvinced, I will happily do a live demo of the problem.

> It has many Mac builds, and its build queue can be quite long at times, and frankly, it is nearly impossible to predict when certain jobs run....

I definitely hear you that we are a big org that makes heavy use of Travis CI. We definitely love the service that you provide. This is exactly why we go to great pains to track down bugs like this one so that we don't generate unneeded builds.

**Comment by jakirkham** [comment]
Please see PR ( https://github.com/conda-forge/pyutilib-feedstock/pull/5 ). This was created 4 days ago. Note it has no outstanding builds as demonstrated by the screenshots below.

<img width="1554" alt="screen shot 2016-10-04 at 9 55 41" src="https://cloud.githubusercontent.com/assets/3019665/19076967/bc7b5fce-8a18-11e6-984a-6467de0264f4.png">

<img width="1554" alt="screen shot 2016-10-04 at 9 54 42" src="https://cloud.githubusercontent.com/assets/3019665/19076935/9ca2a8e2-8a18-11e6-880b-edf8e1bf30d0.png">

**Comment by jakirkham** [comment]
I have now merged PR ( https://github.com/conda-forge/pyutilib-feedstock/pull/5 ). Note that this has triggered a fresh build for the push commit only after merging.

<img width="1554" alt="screen shot 2016-10-04 at 9 58 05" src="https://cloud.githubusercontent.com/assets/3019665/19077050/11a51ef4-8a19-11e6-929d-ccd70de55af0.png">

<img width="1554" alt="screen shot 2016-10-04 at 9 57 28" src="https://cloud.githubusercontent.com/assets/3019665/19077043/05f43cf2-8a19-11e6-94b5-a40f055f2489.png">

**Comment by BanzaiMan** [comment]
https://github.com/conda-forge/poliastro-feedstock/pull/2 and https://github.com/conda-forge/pyutilib-feedstock/pull/5 are a PR into the `master` branch. The merge commit will run on the `master` branch. There is no `feedstock_rerender_master` branch on conda-forge/* repos.

**Comment by jakirkham** [comment]
Yes, they are both PRs into the `master` branch. However, the merge commits for both [`poliastro`]( https://github.com/conda-forge/poliastro-feedstock/commit/af5c402c262c54a27bcde0a1cec634e01ba845b7 ) and [`pyutilib`]( https://github.com/conda-forge/pyutilib-feedstock/commit/de778208554c960ad131daf3b281efa1c295f278 ) says `[ci skip]`, which is being ignored. Instead it is resulting in these older commits being built.

**Comment by jakirkham** [comment]
Could we please have this issue reopened?

**Comment by jakirkham** [comment]
Thanks.

**Comment by jakirkham** [comment]
Also wanted to add one more screenshot to show that the status at merge of PR ( https://github.com/conda-forge/pyutilib-feedstock/pull/5 ) shows it is passing all 4 checks with no pending checks. However, a 5th check shows up on the commit in the PR associated with Travis CI push after the merge. This always shows up after merging a PR like this.

<img width="1554" alt="screen shot 2016-10-07 at 13 00 11" src="https://cloud.githubusercontent.com/assets/3019665/19198610/058873b0-8c8e-11e6-8051-b66c9441ebca.png">

Sorry it is a bit small. Was a bit hard trying to fit everything in. Though one can see this on the PR ( https://github.com/conda-forge/pyutilib-feedstock/pull/5 ) if one wishes to look.

**Comment by jakirkham** [comment]
Should add that if a merge/squash is done this problem doesn't seem to occur. Please see PR ( https://github.com/conda-forge/helper-feedstock/pull/3 ) as an example and the screenshot from it below.

<img width="1554" alt="screen shot 2016-10-07 at 13 45 15" src="https://cloud.githubusercontent.com/assets/3019665/19199907/5cbf2dee-8c94-11e6-9536-aa372f8b2ef3.png">

**Comment by jakirkham** [comment]
Given the backlog issues, maybe revisiting some of the CI skip issues like this one would help.

**Comment by jakirkham** [comment]
Is the code to determine what to build open sourced? Maybe I could take a look at what it would take to fix this.

**Comment by jakirkham** [comment]
> Is the code to determine what to build open sourced?

Any thoughts on this question, @BanzaiMan? This is increasingly a pain point for our org and it would be really great if we could do something to fix it.

**Comment by BanzaiMan** [comment]
@jakirkham That code is not open sourced.

**Comment by BanzaiMan** [comment]
Did I miss any commit that did have `[ci skip]` command in the commit subject and triggered a build?

Take a look at https://github.com/BanzaiMan/travis_production_test/pull/58. When I merged, instead of the default message, I modified the message:

![_ci_skip__empty_commit_by_banzaiman_ _pull_request__58_ _banzaiman_travis_production_test](https://cloud.githubusercontent.com/assets/25666/23717831/64b1cb88-0403-11e7-9866-0852b133d02a.png)

The resulting commit, then, was correctly skipped:

![banzaiman_travis_production_test_-_travis_ci](https://cloud.githubusercontent.com/assets/25666/23717941/cf8c3970-0403-11e7-80ff-b4d79ca98524.png)

**Comment by jakirkham** [comment]
> @jakirkham That code is not open sourced.

Ok, thanks for clarifying.

> Did I miss any commit that did have [ci skip] command in the commit subject and triggered a build?

No, we only added it in the title and not the commit message. This is intentional as we wish to see the build will work in the PR, but know we don't need the build in `master`. If you try this, you should see similar behavior. Please let me know if you need further clarification.

**Comment by jakirkham** [comment]
FWIW put PR ( https://github.com/BanzaiMan/travis_production_test/pull/60 ) up for merging. This should provide the same behavior we see in our org when you merge it. Please make sure to look closely at the status of that commit before and after merging it.

**Comment by Rochet2** [comment]
I believe I have the same problem. The problem I have is that when I create a merge with
`git merge -m "merge commit message [skip ci]" somebranch`
and push the result then travis will build the commit before the merge commit.
You can see this better with an image here:
![Image of the issue in github](https://i.imgur.com/9QiVeS8.png)

Here is an example. This travis job made a merge and push: https://travis-ci.com/Rochet2/TrinityCore3.3.5/jobs/157766489
and it triggered the following build even though I expected it not to create travis builds: https://travis-ci.org/Rochet2/TrinityCore/builds/453672556

**Comment by BanzaiMan** [comment]
https://github.com/travis-ci/travis-ci/issues/553

**Comment by Rochet2** [comment]
Related issue at https://travis-ci.community/t/ci-skip-in-merge-commit-builds-previous-commit/1695

**Comment by BanzaiMan** [comment]
We've pushed the fix now. 

https://changelog.travis-ci.com/skip-ci-command-is-respected-under-all-circumstances-133546
